also the build failure looks real:

```
======================================================================
ERROR: test_authenticate_as (corehq.apps.hqadmin.tests.test_authenticate_as.AuthenticateAsIntegrationTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/hqadmin/tests/test_authenticate_as.py", line 58, in test_authenticate_as
    mobile_worker = CommCareUser.create('potter', 'harry', '123')
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/users/models.py", line 1405, in create
    commcare_user.save(**get_safe_write_kwargs())
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/users/models.py", line 1355, in save
    super(CommCareUser, self).save(**params)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/users/models.py", line 1227, in save
    raise self.Inconsistent("CouchUser with username %s already exists" % self.username)
Inconsistent: CouchUser with username harry already exists
```

--- a/corehq/apps/hqadmin/forms.py
+++ b/corehq/apps/hqadmin/forms.py
@@ -4,7 +4,7 @@ from crispy_forms import layout as crispy
 from django import forms
 from django.core.exceptions import ValidationError
 
-from corehq.apps.users.models import User
+from corehq.apps.users.models import CommCareUser
 
 
 class EmailForm(forms.Form):
@@ -34,20 +34,21 @@ class AuthenticateAsForm(forms.Form):
         domain = self.cleaned_data['domain']
 
         # Ensure that the username exists either as the raw input or with fully qualified name
-        try:
-            User.objects.get(username=username)
-        except User.DoesNotExist:
-            try:
-                extended_username = u"{}@{}.commcarehq.org".format(username, domain)
-                User.objects.get(username=extended_username)
-                self.cleaned_data['username'] = extended_username
-            except User.DoesNotExist:
-                if domain:
-                    raise forms.ValidationError(
-                        u"Cannot find user '{}' for domain '{}'".format(username, domain)
-                    )
-                else:
-                    raise forms.ValidationError(u"Cannot find user '{}'".format(username))
+        if domain:
+            extended_username = u"{}@{}.commcarehq.org".format(username, domain)
+            user = CommCareUser.get_by_username(username=extended_username)
+            self.cleaned_data['username'] = extended_username
+            if user is None:
+                raise forms.ValidationError(
+                    u"Cannot find user '{}' for domain '{}'".format(username, domain)
+                )
+        else:
+            user = CommCareUser.get_by_username(username=username)
+            if user is None:
+                raise forms.ValidationError(u"Cannot find user '{}'".format(username))
+
+        if not user.is_commcare_user():
+            raise forms.ValidationError(u"User '{}' is not a CommCareUser".format(username))
 
         return self.cleaned_data
 

--- a/corehq/apps/hqadmin/forms.py
+++ b/corehq/apps/hqadmin/forms.py
@@ -1,7 +1,6 @@
 import re
 from crispy_forms.helper import FormHelper
 from crispy_forms import layout as crispy
-from crispy_forms.bootstrap import FormActions
 from django import forms
 from django.core.exceptions import ValidationError
 

--- a/corehq/apps/hqadmin/tests/test_authenticate_as.py
+++ b/corehq/apps/hqadmin/tests/test_authenticate_as.py
@@ -11,11 +11,13 @@ class AuthenticateAsFormTest(TestCase):
 
     @classmethod
     def setUpClass(cls):
-        cls.mobile_worker = CommCareUser.create('potter', 'harry', '123')
+        cls.mobile_worker = CommCareUser.create('potter', 'harry@potter.commcarehq.org', '123')
+        cls.regular = WebUser.create('potter', 'awebuser', '***', is_active=True)
 
     @classmethod
     def tearDownClass(cls):
         cls.mobile_worker.delete()
+        cls.regular.delete()
 
     def test_valid_data(self):
         data = {
@@ -28,6 +30,7 @@ class AuthenticateAsFormTest(TestCase):
         data = {
             'username': 'harry@potter.commcarehq.org',
         }
+        form = AuthenticateAsForm(data)
         self.assertTrue(form.is_valid(), form.errors)
 
     def test_invalid_data(self):
@@ -44,6 +47,14 @@ class AuthenticateAsFormTest(TestCase):
         form = AuthenticateAsForm(data)
         self.assertFalse(form.is_valid())
 
+    def test_no_login_as_other_webuser(self):
+        data = {
+            'username': 'awebuser'
+        }
+        form = AuthenticateAsForm(data)
+        self.assertFalse(form.is_valid())
+        self.assertTrue('not a CommCareUser' in str(form.errors))
+
 
 class AuthenticateAsIntegrationTest(TestCase):
 
@@ -57,7 +68,7 @@ class AuthenticateAsIntegrationTest(TestCase):
 
         cls.user.is_superuser = True
         cls.user.save()
-        cls.mobile_worker = CommCareUser.create('potter', 'harry', '123')
+        cls.mobile_worker = CommCareUser.create('potter', 'harry@potter.commcarehq.org', '123')
         cls.regular = WebUser.create(cls.domain, cls.regular_name, cls.password, is_active=True)
 
     @classmethod

--- a/corehq/apps/hqadmin/tests/test_authenticate_as.py
+++ b/corehq/apps/hqadmin/tests/test_authenticate_as.py
@@ -1,4 +1,4 @@
-from django.test import TestCase, SimpleTestCase
+from django.test import TestCase
 from django.core.urlresolvers import reverse
 
 from corehq.apps.users.models import WebUser, CommCareUser
@@ -9,9 +9,15 @@ from ..views import AuthenticateAs
 
 class AuthenticateAsFormTest(TestCase):
 
-    def test_valid_data(self):
-        mobile_worker = CommCareUser.create('potter', 'harry', '123')
+    @classmethod
+    def setUpClass(cls):
+        cls.mobile_worker = CommCareUser.create('potter', 'harry', '123')
 
+    @classmethod
+    def tearDownClass(cls):
+        cls.mobile_worker.delete()
+
+    def test_valid_data(self):
         data = {
             'username': 'harry',
             'domain': 'potter'
@@ -26,14 +32,14 @@ class AuthenticateAsFormTest(TestCase):
 
     def test_invalid_data(self):
         data = {
-            'username': 'harry',
+            'username': 'ron-weasley',
             'domain': 'potter'
         }
         form = AuthenticateAsForm(data)
         self.assertFalse(form.is_valid())
 
         data = {
-            'username': 'harry',
+            'username': 'ron-weasley',
         }
         form = AuthenticateAsForm(data)
         self.assertFalse(form.is_valid())
@@ -45,17 +51,22 @@ class AuthenticateAsIntegrationTest(TestCase):
     def setUpClass(cls):
         cls.domain = 'my-test-views'
         cls.username = 'cornelius'
+        cls.regular_name = 'ron'
         cls.password = 'fudge'
         cls.user = WebUser.create(cls.domain, cls.username, cls.password, is_active=True)
+
         cls.user.is_superuser = True
         cls.user.save()
+        cls.mobile_worker = CommCareUser.create('potter', 'harry', '123')
+        cls.regular = WebUser.create(cls.domain, cls.regular_name, cls.password, is_active=True)
 
     @classmethod
     def tearDownClass(cls):
         cls.user.delete()
+        cls.mobile_worker.delete()
+        cls.regular.delete()
 
     def test_authenticate_as(self):
-        mobile_worker = CommCareUser.create('potter', 'harry', '123')
         self.client.login(username=self.username, password=self.password)
 
         resp = self.client.get(reverse(AuthenticateAs.urlname))
@@ -67,5 +78,20 @@ class AuthenticateAsIntegrationTest(TestCase):
         form.data['domain'] = 'potter'
 
         resp = self.client.post(reverse(AuthenticateAs.urlname), form.data, follow=True)
-        self.assertEqual(resp.status_code, 200)
-        self.assertEqual(resp.context['user'].username, 'harry')
+        self.assertEqual(
+            self.client.session['_auth_user_id'],
+            self.mobile_worker.get_django_user().id
+        )
+
+    def test_permisssions_for_authenticate_as(self):
+        self.client.login(username=self.regular_name, password=self.password)
+
+        resp = self.client.get(reverse(AuthenticateAs.urlname))
+        self.assertTrue('no_permissions' in resp.url)
+
+        resp = self.client.post(reverse(AuthenticateAs.urlname), {})
+        self.assertTrue('no_permissions' in resp.url)
+        self.assertEqual(
+            self.client.session['_auth_user_id'],
+            self.regular.get_django_user().id
+        )

--- a/settings.py
+++ b/settings.py
@@ -679,6 +679,7 @@ AUDIT_MODEL_SAVE = [
 
 AUDIT_VIEWS = [
     'corehq.apps.settings.views.ChangeMyPasswordView',
+    'corehq.apps.hqadmin.views.AuthenticateAs',
 ]
 
 AUDIT_MODULES = [

